src/pip/_internal/req/req_install.py

```diff
@@ -314,6 +314,7 @@ def from_path(self) -> Optional[str]:
             return None
         s = str(self.req)
         if self.comes_from:
+            comes_from: Optional[str]
```

Comment:
Why is this needed?

Reply:
mypy's inference isn't perfect. It infers a type of `str` for `comes_from`, but then complains when it notices that `self.comes_from.from_path()` can return None. Adding the explicit type annotation helps mypy out.